### PR TITLE
implement per-container apparmor profiles

### LINKED ISSUE #210: Add per-container apparmor profile generation
Generate an apparmor profile per container, based on a reasonable default profile + whatever the container set in raw.apparmor (if anything).

### PR DESCRIPTION
**Pull request by tych0**
Closes #210

Signed-off-by: Tycho Andersen <tycho.andersen@canonical.com>

### CODE REVIEW

File: lxd/apparmor.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	 * Since it uses mtimes, we shouldn't just always write out our local
+	 * apparmor template; instead we should check to see whether the
+	 * template is the same as ours. If it isn't we should write our
+	 * version out so that the new changes are reflected and we definitely
+	 * force a recompile.
+	 */
+	profile := path.Join(aaPath, "profiles", AAProfileName(c))
+	content, err := ioutil.ReadFile(profile)
+	if err != nil && !os.IsNotExist(err) {
+		shared.Log.Error("TYCHO: readfile failed")
```
Comment (stgraber) [inline lxd/apparmor.go]: leftover debugging

File: lxd/apparmor.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	rawApparmor, ok := c.config["raw.apparmor"]
+	if !ok {
+		rawApparmor = ""
+	}
+
+	return fmt.Sprintf(DEFAULT_POLICY, c.name, rawApparmor)
+}
+
+func runApparmor(command string, profile string) error {
+	cmd := exec.Command("apparmor_parser", []string{
```
Comment (stgraber) [inline lxd/apparmor.go]: We should probably check for apparmor_parser's presence and if missing, not mess with lxc.aa_profile (for distros that don't ship with apparmor).

Comment (stgraber) [inline lxd/apparmor.go]: That check should ideally happen at daemon startup time and print a log entry if apparmor support is disabled.

File: lxd/apparmor.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+	updated := getProfileContent(c)
+
+	if string(content) != string(updated) {
+		if err := os.MkdirAll(path.Join(aaPath, "profiles"), 0700); err != nil {
+			return err
+		}
+
+		if err := ioutil.WriteFile(profile, []byte(updated), 0600); err != nil {
+			shared.Log.Error("TYCHO: write file failed")
```
Comment (stgraber) [inline lxd/apparmor.go]: more leftover debugging

**Comment by stgraber** [comment]
One thing we can technically do and which may not be too difficult to implement (just an extra load call) is for the apparmor profile to be updated immediately when raw.apparmor is changed.

**Comment by tych0** [comment]
On Wed, Sep 02, 2015 at 09:16:38PM -0700, Stéphane Graber wrote:
> One thing we can technically do and which may not be too difficult to implement (just an extra load call) is for the apparmor profile to be updated immediately when raw.apparmor is changed.

Ah, yeah. I can do that.

**Comment by stgraber** [comment]
Another thing I'm wondering is whether "raw.lxc: lxc.aa_profile=unconfined" still works with this change?
We don't yet have the knobs needed to do this cleanly so having the workaround way to still work would be nice.

**Comment by tych0** [comment]
On Wed, Sep 02, 2015 at 09:18:32PM -0700, Stéphane Graber wrote:
> Another thing I'm wondering is whether "raw.lxc: lxc.aa_profile=unconfined" still works with this change?
> We don't yet have the knobs needed to do this cleanly so having the workaround way to still work would be nice.

It should, although the way it's currently written setting aa_profile
on the container will work, but setting it on a profile won't. I'll
move it up so that it is processed before profiles too.

**Pull request merged**